### ISSUE TITLE
Rating Limiting on DELETE requests is absurdly restrictive

### ISSUE THREAD

**Issue by generaltso** [tootsuite/mastodon#10557]
I want to delete all my statuses, every inane comment, every reply, all of it.

**I do not want to delete my account.** From this point forward I plan to delete everything older than 1-3 days every day.

I would prefer not to have an eternally preserved transcript of every message I have ever posted on Mastodon or anywhere else online for that matter. I would prefer for my messages to be ephemeral. I have the right to be forgotten.

Right now, statuses must be deleted with individual DELETE requests and are rate-limited at 30 requests every 30 minutes. Glancing at the source, this is a hard-coded limit. Using either the Web UI or the API results in the same restriction.

Effectively, one can only delete 1 toot per minute. As I write this issue, my script deleting 1500 messages will complete its task in 24 hours from now. If I had 15,000 that would be 10 days. This is absurd. If I had database access this would be a single operation.

Please consider doing one of the following:

1) Allow multiple statuses to be deleted in a single request
2) Increase the rate limit to something more reasonable (optionally make it customize-able per-instance) 
3) Add a purge feature somewhere to let users choose to delete posts older/newer than some amount of time

**Comment by trwnh** [comment]
>If I had database access this would be a single operation.

That wouldn't send out a Delete to other servers. The limiting factor here isn't *your* server, it's *other* servers. Another thing to watch out for is that each status deleted via the API also triggers an individual task in the Sidekiq queue, so it can make performance worse for all other users of the same server. See #9384 for an explanation of this.

On a local level, this can be improved. On a federation level, I'm not sure if there's any way around how expensive it is to be sending out a bunch of Delete activities in too short of a timeframe.

**Comment by nsuchy** [comment]
> > If I had database access this would be a single operation.
> 
> That wouldn't send out a Delete to other servers. The limiting factor here isn't _your_ server, it's _other_ servers. Another thing to watch out for is that each status deleted via the API also triggers an individual task in the Sidekiq queue, so it can make performance worse for all other users of the same server. See #9384 for an explanation of this.
> 
> On a local level, this can be improved. On a federation level, I'm not sure if there's any way around how expensive it is to be sending out a bunch of Delete activities in too short of a timeframe.

Why not allow server owners to allow users to queue their deletes? Knowing this is a federated network, users should understand that any delete or edit operation could take up to a few days for all servers to respect the change anyways and rouge servers can choose to ignore it.

**Comment by lapineige** [comment]
I've the same issues with the rate limit, that is so low than I wasn't able to delete more than ~20K statuses in 6 months. The current limit is really low, but we might also consider than the tools used to delete the statuses are not always running an a server (that might create some privacy issues), for instance if people use https://github.com/kensanata/mastodon-backup#expiring-your-toots-and-favourites, and as a result can't erase content all day long. As a result the "averaged" rate limit of 1 status/min can in practice become rather less than 0,5/min (assuming the computer is used 12h/day, everyday, which is a lot, and it will probably be a lot smaller).
In my case, I'm closer to 0,1 status/min, 1 status each 10min.
For some active accounts, 1 toot/boost each 10min can be really low.
So there an issue where you can't actually delete all your statuses as fast as your creating new one (and it takes *forever* to delete a bunch of old toots…).

Could there be another way to manage those deletion ? I don't know the technical part, but would it be possible to queue the request, and only delete the content when the server charge is low ?
For instance running the deletion during the night, as most servers connected are probably in a similar timezone (I guess ?).

**Comment by leaxael** [comment]
Despite the tag of this thread, this is apparently [not partially a bug but a design choice](https://mastodon.social/@Gargron/101588449409740014).

I'm curious to know if this limitation is justifiable by backend performance concerns, i.e. if there's any data/observation we can refer to, rather than simply claiming the conclusion. 1toot/1minute deletion is a significant restriction even for _manual_ deletions. Yes it takes some time to propagate through the federated network, but I believe that is a different process to be handled by the housekeeping parts, instead of by a API rate constraint. Users should be allowed to simply submit the requests. 

BTW, [the documentation](https://docs.joinmastodon.org/api/rate-limits/) does not specify how long would the 429 error cost but it seems to be 20-30minutes, _shared by all tokens/applications of the account_.

**Comment by trwnh** [comment]
it's still "partially a bug" in that it could be fixed/mitigated with the recommendations in the original comment (e.g. batching Deletes). the "design choice" is based solely on how expensive deletes are right now. the API rate limit is to prevent client apps from spamming Deletes and overloading their own server (and others). you simply don't want your sidekiq queues to grow so large that other functionality is affected by it. as for the docs, they clearly state that the limit is by account, not by token.

**Comment by lapineige** [comment]
> Users should be allowed to simply submit the requests.

I would second that regarding the impact on use case : having this 30min limit means you can't delete a lot of toots except if you send removal request at least each 30min.
Which means the only way to delete some toots from "big" account will take a long time, but it would take forever (= it is not possible) without using a web service to do so (you won't lunch you deletion tool each 30min, and you can only do that during day time, and so on).
That's a big restriction and might encourage some use case that can be privacy-harmful (if you rely on a third party server to access you account for you toot deletion, that creates a privacy risk).

**Comment by leaxael** [comment]
I emphasized the per account limit because 30toots/30minutes is so low that you can accidentally hit the ceiling by manually removing some toots on your phone/browser.

**Comment by leaxael** [comment]
Similar to @lapineige 's comments, I would like to emphasize that not everyone has server(s). My work pc is almost up all the time, but will be down when I move around with it. With the 30/30 API rule it would be still very difficult for me to crontab a script, as the process has to be alive / resumed otherwise to send requests every 30 minutes, instead of being done within a short-ish timeframe. To delete 10k toots, it's 166 hours or 7 full days without halt.

Also, the 30/30 seems to count failed delete requests too (i.e. toot is already removed), which means to maintain the query queue on the user's side, there's more to keep track of than just a list of IDs and the 30 minute interval.

**Comment by trwnh** [comment]
> the current mastodon/fediverse has a few huge instances hosting the majority of the user base

it also has thousands of instances hosting a not-insignificant amount of users. you have to think about the "long tail" in this case.

i do agree that the rate limit is low, and i think that the federation cost can be reduced if activitypub implementations cooperate and support batched deletes, for example. the spec doesn't prevent listing an array for the `object` of a `Delete` activity, so it should be theoretically possible to send out multi-status delete activities. in practice, that would require consensus and the propagation of software updates containing such a patch. and that would still leave the local cost of delete tasks being added to the sidekiq job queue

**Comment by leaxael** [comment]
Yeah I feel like deleting in batch should be possible, but until now I haven't looked into the internals of mastodon or other implementations. Maybe I should try.

> you have to think about the "long tail" in this case.

Sorry for dragging this digress, but that's exactly what I tried to express - request propagation's complexity will grow above linear, 30/30 might cope for the moment but perhaps not in the future, where we _hope_ the fedi network's graph to be more connected than sparse, i.e. more smaller to medium-sized instances here and there.

**Comment by KellyHang** [comment]
www
